### PR TITLE
Replace custom ChaCha20 cipher with RustCrypto ChaCha20 implementation

### PR DESCRIPTION
**Pull request by philipr-za**
## Description
Fuzzing revealed that the custom ChaCha20 implementation in `tari_utilities` was not correct so this PR migrates the DHT encryption to use the audited RustCrypto ChaCha20 encryption. 

Had to bump the toolchain version for this update so updated to the most current version of nightly. The only update needed was adding a new feature flag #![feature(min_type_alias_impl_trait)] which is a new flag for the same issue we already use (i.e. #![feature(type_alias_impl_trait)])

The new nightly toolchain contained an update to the linting used by rustfmt so I have included a second commit that is just new formatting to make this easier to review.

## How Has This Been Tested?
Tests updated

## Checklist:

* [ ] I'm merging against the `development` branch.
* [ ] I have squashed my commits into a single commit.

### CODE REVIEW

**Review by stringhandler — APPROVED** [review]

File: comms/dht/src/crypt.rs [hunk 1]
```diff
@@ -20,29 +20,55 @@
...
+use std::mem::size_of;
 
 pub fn generate_ecdh_secret<PK>(secret_key: &PK::K, public_key: &PK) -> PK
 where PK: PublicKey + DiffieHellmanSharedSecret<PK = PK> {
     PK::shared_secret(secret_key, public_key)
 }
 
-pub fn decrypt(cipher_key: &CommsPublicKey, cipher_text: &[u8]) -> Result<Vec<u8>, CipherError> {
-    ChaCha20::open_with_integral_nonce(&cipher_text.to_vec(), cipher_key.as_bytes())
+pub fn decrypt(cipher_key: &CommsPublicKey, cipher_text: &[u8]) -> Result<Vec<u8>, DhtOutboundError> {
```
Comment (CjS77) [inline comms/dht/src/crypt.rs:43]: Is decypting with a pubkey correct? How is the en- / decrpytion used in the comms? If it's all with pubkeys, then the encryption is likely meaningless.

File: comms/dht/src/crypt.rs [hunk 2]
```diff
@@ -63,10 +89,10 @@ mod test {
     fn decrypt_fn() {
         let key = CommsPublicKey::default();
         let cipher_text =
-            from_hex("7ecafb4c0a88325c984517fca1c529b3083e9976290a50c43ff90b2ccb361aeaabfaf680e744b96fc3649a447b")
+            from_hex("24bf9e698e14938e93c09e432274af7c143f8fb831f344f244ef02ca78a07ddc28b46fec536a0ca5c04737a604")
```
Comment (CjS77) [inline comms/dht/src/crypt.rs:99]: This is odd, I thought only edge cases were broken on the handrolled chacha. If so, then this should still work with the same input and this suggests that the API has changed significantly, meaning other parts of comms could break.

File: comms/src/lib.rs [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #![recursion_limit = "512"]
 // Allow `type Future = impl Future`
 #![feature(type_alias_impl_trait)]
+#![feature(min_type_alias_impl_trait)]
```
Comment (CjS77) [inline comms/src/lib.rs:20]: This feature [may not be stabilised for a while](https://github.com/rust-lang/rust/issues/63063#issuecomment-795682305) so you might be hampering the effort to move Tari to stable Rust.

**Review by CjS77 — COMMENTED** [review]
A quick overview. I have a few comments and questions.

File: comms/dht/src/crypt.rs (hunk 1 above)
Comment (philipr-za) [reply comms/dht/src/crypt.rs:43]: Its a Diffie-Hellman key

File: comms/src/lib.rs (hunk 3 above)
Comment (philipr-za) [reply comms/src/lib.rs:20]: It is the same issue the feature above which is critical to large portions of the code, they share the same tracking Issue in the Rust codebase.

File: comms/dht/src/crypt.rs (hunk 1 above)
Comment (stringhandler) [reply comms/dht/src/crypt.rs:43]: The type here is confusing. Perhaps just a different type?

Comment (CjS77) [reply comms/dht/src/crypt.rs:43]: DH keys are typically secret keys. Yeah, it's just confusing.

Comment (philipr-za) [reply comms/dht/src/crypt.rs:43]: Would need to update the way its implemented in the `tari_crypto` crate. 
https://github.com/tari-project/tari-crypto/blob/main/src/keys.rs#L78

Comment (CjS77) [reply comms/dht/src/crypt.rs:43]: Yes, it's incorrect in tari_crypto. @hansieodendaal picked that up a few weeks ago and was going to push a PR changing it.

File: comms/dht/src/crypt.rs (hunk 2 above)
Comment (philipr-za) [reply comms/dht/src/crypt.rs:99]: So looked further into this and I cannot actually get the RustCrypto output to agree with the custom implementation or with the test vectors defined in https://datatracker.ietf.org/doc/html/rfc8439#section-2.4.1. However, I checked the output against the Python Cryptodome library and the RustCrypto and Cryptodome implementations agree 🤷‍♂️

**Review by stringhandler — APPROVED** [review]

**Comment by philipr-za** [comment]
Updating the rust toolchain in this PR resulted in the rustfmt and clippy being updated so most of the PR are those updates. Just for clarity the only file that was actually updated in terms of logic is noted below.

File: comms/dht/src/crypt.rs [hunk 4]
```diff
@@ -20,29 +20,62 @@
...
 use tari_crypto::{
     keys::{DiffieHellmanSharedSecret, PublicKey},
-    tari_utilities::{
-        ciphers::{
-            chacha20::ChaCha20,
-            cipher::{Cipher, CipherError},
-        },
-        ByteArray,
-    },
+    tari_utilities::ByteArray,
```
Comment (philipr-za) [inline comms/dht/src/crypt.rs:35]: This is the file that was modified in this PR.

**Pull request merged**